amethyst_rendy/src/plugins.rs

```diff
@@ -0,0 +1,322 @@
+                self.dirty = true;
+                self.dimensions = new_dimensions.map(|d| d.clone());
+                return false;
+            }
+            self.dirty
+        }
+
+        fn on_plan(
+            &mut self,
+            plan: &mut RenderPlan<B>,
+            factory: &mut Factory<B>,
+            res: &Resources,
+        ) -> Result<(), Error> {
+            self.dirty = false;
+
+            let window = <ReadExpect<'_, Window>>::fetch(res);
+            let surface = factory.create_surface(&window);
+            let dimensions = self.dimensions.as_ref().unwrap();
+            let window_kind = Kind::D2(dimensions.width() as u32, dimensions.height() as u32, 1, 1);
+
+            let depth_options = ImageOptions {
+                kind: window_kind,
+                levels: 1,
+                format: Format::D32Sfloat,
+                clear: Some(ClearValue::DepthStencil(ClearDepthStencil(1.0, 0))),
+            };
+
+            plan.add_root(Target::Main);
+            plan.define_pass(
+                self.target,
+                crate::bundle::TargetPlanOutputs {
+                    colors: vec![OutputColor::Surface(
+                        surface,
+                        self.clear.map(ClearValue::Color),
+                    )],
+                    depth: Some(depth_options),
+                },
+            )?;
+
+            Ok(())
+        }
+    }
+}
+
+/// A `RenderPlugin` for forward rendering of 3d objects using flat shading.
+pub type RenderFlat3D = RenderBase3D<crate::pass::FlatPassDef>;
+/// A `RenderPlugin` for forward rendering of 3d objects using shaded shading.
+pub type RenderShaded3D = RenderBase3D<crate::pass::ShadedPassDef>;
+/// A `RenderPlugin` for forward rendering of 3d objects using physically-based shading.
+pub type RenderPbr3D = RenderBase3D<crate::pass::PbrPassDef>;
+
+/// A `RenderPlugin` for forward rendering of 3d objects.
+/// Generic over 3d pass rendering method.
+#[derive(derivative::Derivative)]
+#[derivative(Default(bound = ""), Debug(bound = ""))]
+pub struct RenderBase3D<D: Base3DPassDef> {
+    target: Target,
+    skinning: bool,
+    marker: std::marker::PhantomData<D>,
+}
+
+impl<D: Base3DPassDef> RenderBase3D<D> {
+    /// Set target to which 3d meshes will be rendered.
+    pub fn with_target(mut self, target: Target) -> Self {
+        self.target = target;
+        self
+    }
+
+    /// Enable rendering for skinned meshes.
+    ///
+    /// NOTE: You must register `VertexSkinningBundle` yourself.
+    pub fn with_skinning(mut self) -> Self {
+        self.skinning = true;
+        self
+    }
+}
+
+impl<B: Backend, D: Base3DPassDef> RenderPlugin<B> for RenderBase3D<D> {
+    fn on_build<'a, 'b>(&mut self, builder: &mut DispatcherBuilder<'a, 'b>) -> Result<(), Error> {
+        builder.add(VisibilitySortingSystem::new(), "visibility_system", &[]);
+
+        // TODO: We should ideally register `VertexSkinningBundle` here,
```

Comment:
Wouldn't it make sense to then move this renderplugin to the animation bundle? If it can only be used with the `VertexSkinningBundle`, it would be sensible to only allow importing them together.

Might also be a good use case for a glue crate?

Reply:
Not really. You can still use that plugin without vertex skinning. It's optional feature, you need it only when `self.skinning` is `true`. Also that bundle has it's own set of configuration that you would have to proxy through that call. Having a NOTE in docs is probably enough.

Reply:
I'll actually remove that TODO as it's not actionable in any way.

Reply:
I see, fair enough